### PR TITLE
New datasets and metrics

### PR DESCRIPTION
**Pull request by albanPi**
## Modifications
- Added new metrics (balanced accuracy and ROC AUC).
- Added new datasets from [Leo's paper](https://hal.science/hal-03723551v2) and package requirements for these datasets.
- Added iris dataset from scikit-learn.
- Modified `objective.py` to add `train_test_split` in `set_data` method.
- Removed `train_test_split` in `simulated.py` to match with `objective.py` changes.

### CODE REVIEW

File: datasets/datasets_openml.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+class Dataset(BaseDataset):
+
+    name = 'openml'
+
+    install_cmd = 'conda'
+    requirements = ["pip:openml", "pip:chardet"]
+
+    parameters = {
+        "test_size": [0.25],
+        "seed": [42],
```
Comment (tomMoral) [inline datasets/datasets_openml.py]: These parameters are not used so could you remove them?

File: datasets/iris.py [hunk 2]
```diff
@@ -16,16 +14,13 @@ class Dataset(BaseDataset):
     }
```
Comment (tomMoral) [inline datasets/iris.py]: Remove the parameters that are not used.

File: datasets/simulated.py [hunk 3]
```diff
@@ -35,18 +32,14 @@ def get_data(self):
         # API to pass data. It is customizable for each benchmark.
```
Comment (tomMoral) [inline datasets/simulated.py]: Remove `test_size` parameter that is not used.

File: benchmark_utils/gridsearch_solver.py [hunk 4]
```diff
@@ -1,59 +1,62 @@
...
+        preprocessor = ColumnTransformer(
+                    [("one_hot", OHE(categories="auto",
+                                     handle_unknown="ignore"),
+                     [i for i in range(size) if self.cat_ind[i]]),
+                     (
+                        "numerical",
+                        "passthrough",
+                        [i for i in range(size) if not self.cat_ind[i]],
+                    )]
+                )
```
Comment (tomMoral) [inline benchmark_utils/gridsearch_solver.py]:
```suggestion
            [("one_hot", OHE(categories="auto",
                             handle_unknown="ignore"),
             [i for i in range(size) if self.cat_ind[i]]),
             (
                "numerical",
                "passthrough",
                [i for i in range(size) if not self.cat_ind[i]],
            )]
        )

```

File: benchmark_utils/gridsearch_solver.py [hunk 5]
```diff
@@ -1,59 +1,62 @@
...
+
+
+# The benchmark solvers must be named `Solver` and
+# inherit from `BaseSolver` for `benchopt` to work properly.
+class GSSolver(BaseSolver):
+
+    stopping_criterion = SufficientProgressCriterion(strategy='callback')
+
+    def set_objective(
+            self, X_train, y_train, X_test, y_test,
```
Comment (tomMoral) [inline benchmark_utils/gridsearch_solver.py]: Do not pass `X_test, y_test` to the classifier, it should not access it.

**Review by tomMoral — COMMENTED** [review]
Some comments

File: datasets/datasets_openml.py (hunk 1 above)
Comment (albanPi) [reply datasets/datasets_openml.py]: I've removed them now, I kept them at first because I got a error saying that these parameters were missing (this concerns the three datasets).

File: benchmark_utils/gridsearch_solver.py (hunk 5 above)
Comment (albanPi) [reply benchmark_utils/gridsearch_solver.py]: If you talk about the arguments, I can't remove them otherwise I'm going to have an error.

Comment (tomMoral) [reply benchmark_utils/gridsearch_solver.py]: you can remove them if you also remove them from `get_objective` or am I missing something?

Comment (albanPi) [reply benchmark_utils/gridsearch_solver.py]: I need `X_test` and `y_test` for `optuna_solver.py` so I can't remove them from `get_objective`.

File: datasets/simulated.py [hunk 6]
```diff
@@ -35,18 +30,14 @@ def get_data(self):
         # API to pass data. It is customizable for each benchmark.
 
         # Generate pseudorandom data using `numpy`.
-        rng = np.random.RandomState(self.seed)
         X, y, _ = make_correlated_data(self.n_samples, self.n_features)
```
Comment (tomMoral) [inline datasets/simulated.py]: you need to keep the seed to make this dataset reproducible here

File: benchmark_utils/optuna_solver.py [hunk 7]
```diff
@@ -20,16 +22,20 @@ class OSolver(BaseSolver):
...
     ):
         # Define the information received by each solver from the objective.
         # The arguments of this function are the results of the
         # `Objective.get_objective`. This defines the benchmark's API for
         # passing the objective to the solver.
         # It is customizable for each benchmark.
-        self.X_train, self.y_train = X_train, y_train
-        self.X_test, self.y_test = X_test, y_test
+        rng = np.random.RandomState(42)
+        X, X_val, y, y_val = train_test_split(
```
Comment (tomMoral) [inline benchmark_utils/optuna_solver.py:37]: you need to put as a parameter of the class

File: benchmark_utils/gridsearch_solver.py [hunk 8]
```diff
@@ -1,59 +1,63 @@
...
+    from sklearn.pipeline import Pipeline
+    from sklearn.compose import ColumnTransformer
+    from sklearn.preprocessing import OneHotEncoder as OHE
+
+
+# The benchmark solvers must be named `Solver` and
+# inherit from `BaseSolver` for `benchopt` to work properly.
+class GSSolver(BaseSolver):
+
+    stopping_criterion = SufficientProgressCriterion(strategy='callback')
```
Comment (tomMoral) [inline benchmark_utils/gridsearch_solver.py]: Here, this should be `SingleRunCriterion` no?

File: benchmark_utils/optuna_solver.py [hunk 9]
```diff
@@ -19,17 +21,27 @@ class OSolver(BaseSolver):
...
         # It is customizable for each benchmark.
-        self.X_train, self.y_train = X_train, y_train
-        self.X_test, self.y_test = X_test, y_test
+        rng = np.random.RandomState(self.params['seed'])
+        X, X_val, y, y_val = train_test_split(
+            X_train, y_train, test_size=self.params['test_size'],
+            random_state=rng
+        )
+        self.X_train, self.y_train = X, y
+        self.X_test, self.y_test = X_val, y_val
```
Comment (tomMoral) [inline benchmark_utils/optuna_solver.py]: Can you rename `X_val, y_val`?

File: datasets/simulated.py [hunk 10]
```diff
@@ -5,10 +5,8 @@
 # - skipping import to speed up autocompletion in CLI.
 # - getting requirements info when all dependencies are not installed.
 with safe_import_context() as import_ctx:
-    import numpy as np
-    from sklearn.model_selection import train_test_split
-
     from benchopt.datasets import make_correlated_data
+    import numpy as np
```
Comment (tomMoral) [inline datasets/simulated.py:9]: Import from classical packages are usually put before import from the current package (benchopt).
This is nitpick but help with hierarchical information.

**Review by tomMoral — APPROVED** [review]
Just a few more nitpicks but otherwise LGTM!

File: benchmark_utils/optuna_solver.py [hunk 11]
```diff
@@ -19,17 +21,27 @@ class OSolver(BaseSolver):
...
         # `Objective.get_objective`. This defines the benchmark's API for
         # passing the objective to the solver.
         # It is customizable for each benchmark.
-        self.X_train, self.y_train = X_train, y_train
-        self.X_test, self.y_test = X_test, y_test
+        rng = np.random.RandomState(self.params['seed'])
+        X, X_val, y, y_val = train_test_split(
+            X_train, y_train, test_size=self.params['test_size'],
+            random_state=rng
+        )
```
Comment (tomMoral) [inline benchmark_utils/optuna_solver.py]: A simpler way actually is:
```suggestion
        X, X_val, y, y_val = train_test_split(
            X_train, y_train, test_size=self.params['test_size'],
            random_state=self.params['seed']
        )

```

File: benchmark_utils/optuna_solver.py [hunk 12]
```diff
@@ -5,12 +5,14 @@
 # - skipping import to speed up autocompletion in CLI.
 # - getting requirements info when all dependencies are not installed.
 with safe_import_context() as import_ctx:
+    import numpy as np
```
Comment (tomMoral) [inline benchmark_utils/optuna_solver.py]:
```suggestion
```

**Review by tomMoral — APPROVED** [review]
A last nitpick :)
